common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,408 @@
+
+/**
+ * A cache that allows multiple readers, but only a single writer, at
+ * a time. It will make a best effort attempt to coalesce reads, but
+ * does not guarantee that all overlapping reads will be coalesced.
+ *
+ * The cache operates by bracketing all reads and writes. A read
+ * imposes a lightweight read lock, by inserting an entry into the
+ * cache with State.ReadInProgress. A write does the same, with
+ * State.WriteInProgress.
+ *
+ * On read or write completion, the value transitions to State.Cached.
+ *
+ * State.Initial is represented implicitly by absence from the cache.
+ *
+ * The handshake for cache entry state transition is:
+ *
+ * 1. if entry is in an agreeable state, proceed
+ *      where agreeable for reads is Initial, ReadInProgress, or Cached
+ *                    and the read proceeds by ensuring the entry is State.ReadInProgress
+ *      and agreeable for writes is Initial or Cached
+ *                    and the write proceeds by ensuring the entry is State.WriteInProgress
+ *      and agreeable for deletes is Cached
+ *
+ * 2. if entry is not in an agreeable state, then read- or write-around the cache;
+ *    for deletions, we allow the delete to proceed, and mark the entry as InvalidateWhenDone
+ *    the owning reader or writer is then responsible for ensuring that the entry is invalid
+ *    when that read or write completes
+ *
+ * 3. to swap in the new state to an existing entry, we use an AtomicReference.compareAndSet
+ *
+ * 4. only if the db operation completes with success, atomically set the state to Cached.
+ *
+ * 5. lastly, for cache invalidations that race with, we mark the entry as
+ *
+ */
+private object MultipleReadersSingleWriterCache {
+    /** Each entry has a state, as explained in the class comment above. */
+    object State extends Enumeration {
+        type State = Value
+        val ReadInProgress, WriteInProgress, InvalidateInProgress, InvalidateWhenDone, Cached = Value
+    }
+
+    import State._
+
+    /** Failure modes, which will only occur if there is a bug in this implementation */
+    case class ConcurrentOperationUnderRead(actualState: State) extends Exception(s"Cache read started, but completion raced with a concurrent operation: $actualState.")
+    case class ConcurrentOperationUnderUpdate(actualState: State) extends Exception(s"Cache update started, but completion raced with a concurrent operation: $actualState.")
+    case class SquashedInvalidation(actualState: State) extends Exception(s"Cache invalidation squashed due $actualState.")
+    case class StaleRead(actualState: State) extends Exception(s"Attempted read of invalid entry due to $actualState.")
+}
+
+trait MultipleReadersSingleWriterCache[W, Winfo] {
+    import MultipleReadersSingleWriterCache._
+    import MultipleReadersSingleWriterCache.State._
+
+    /** Subclasses: Toggle this to enable/disable caching for your entity type. */
+    protected lazy val cacheEnabled = true
```

Comment:
it's used as a default argument in DocumentFactory.
`    def get[Wsuper >: W](db: ArtifactStore[Wsuper], doc: DocId, rev: DocRevision = DocRevision(), fromCache: Boolean = cacheEnabled)`